File: opi-core/src/main/java/org/lei/opi/core/PhoneHMD.java
package org.lei.opi.core;

import java.util.HashMap;

import org.lei.opi.core.structures.Parameter;
import org.lei.opi.core.structures.ReturnMsg;

/**
 * PhoneHMD client
 *
 * @since 0.0.1
 */
public class PhoneHMD extends Jovp {

  /**
   * PhoneHMD constructor
   *
   * @since 0.0.1
   */
  public PhoneHMD() {
    super();
  }

  /**
   * opiQuery: Query device
   * 
   * @return settings and state machine state
   *
   * @since 0.0.1
   */
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error message.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.jovp", desc = "Any messages that the JOVP sent back.")
  @ReturnMsg(name = "msg.isTracking", desc = "0 eye tracking is off, any other value it is on.", className = Double.class, min = 0)
  @ReturnMsg(name = "msg.isCalibrated", desc = "0 eye tracking has not been calibrated, any other value it has.", className = Double.class, min = 0)
  public MessageProcessor.Packet query() {
    return new MessageProcessor.Packet("");
  }

  /**
   * opiInitialise: initialize OPI
   * 
   * @param args A map of name:value pairs for Params
   * 
   * @return A JSON object with return messages
   * 
   * @since 0.0.1
   */
  @Parameter(name = "ip", desc = "IP Address of the perimeter.", defaultValue = "192.126.0.1")
  @Parameter(name = "port", desc = "TCP port of the perimeter.", className = Double.class, min = 0, max = 65535, defaultValue = "50000")
  @Parameter(name = "ip_Monitor", desc = "IP Address of the OPI JOVP server.", defaultValue = "localhost")
  @Parameter(name = "port_Monitor", desc = "TCP port of the OPI JOVP server.", className = Double.class, min = 0, max = 65535, defaultValue = "50001")
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error messages from PhoneHMD.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.jovp", desc = "Any messages that the JOVP sent back.")
  public MessageProcessor.Packet initialize(HashMap<String, Object> args) {
    setIsInitialised(true);
    return new MessageProcessor.Packet("");
  }

  /**
   * opiSetup: Change device background and overall settings
   * 
   * @param args pairs of argument name and value
   * 
   * @return A JSON object with return messages
   *
   * @since 0.0.1
   */
  @Parameter(name = "eye", desc = "Eye to set.", className = Eye.class, defaultValue = "both")
  @Parameter(name = "bgLum", desc = "Background color for eye.", className = Double.class, defaultValue = "10", min = 0, max = 3183.099)
  @Parameter(name = "bgCol", desc = "Background color for eye.", className = Double[].class, isList = true, defaultValue = "list(1, 1, 1)")
  @Parameter(name = "fixType", desc = "Fixation target type for eye.", className = Shape.class, defaultValue = "maltese")
  @Parameter(name = "fixLum", desc = "Fixation target luminance for eye.", className = Double.class, defaultValue = "20", min = 0, max = 3183.099)
  @Parameter(name = "fixCol", desc = "Fixation target color for eye.", className = Double[].class, isList = true, defaultValue = "list(0, 1, 0)")
  @Parameter(name = "fixCx", desc = "x-coordinate of fixation target (degrees).", className = Double.class, min = -90, max = 90, defaultValue = "0")
  @Parameter(name = "fixCy", desc = "y-coordinate of fixation target (degrees).", className = Double.class, min = -90, max = 90, defaultValue = "0")
  @Parameter(name = "fixSx", desc = "diameter along major axis of ellipse (degrees).", className = Double.class, min = 0, max = 180, defaultValue = "1")
  @Parameter(name = "fixSy", desc = "diameter along minor axis of ellipse (degrees).", className = Double.class, min = 0, max = 180, defaultValue = "1")
  @Parameter(name = "fixRotation", desc = "Angles of rotation of fixation target (degrees). Only useful if sx != sy specified.", className = Double.class, min = 0, max = 360, defaultValue = "0")
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error messages from PhoneHMD.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.jovp", desc = "Any messages that the JOVP sent back.")
  public MessageProcessor.Packet setup(HashMap<String, Object> args) {
    return new MessageProcessor.Packet("");
  }

  /**
   * opiPresent: Present OPI stimulus in perimeter
   * 
   * @param args pairs of argument name and value
   * 
   * @return A JSON object with return messages
   *
   * @since 0.0.1
   */
  @Parameter(name = "eye" , desc = "Eye to test.", className = Eye.class, isList = true, defaultValue = "list('left')")
  @Parameter(name = "shape" , desc = "Stimulus shape.", className = Shape.class, isList = true, defaultValue = "list('circle')", optional = true)
  @Parameter(name = "type" , desc = "Stimulus type.", className = Type.class, isList = true, defaultValue = "list('flat')", optional = true)
  @Parameter(name = "x", desc = "List of x co-ordinates of stimuli (degrees).", className = Double.class, min = -90, max = 90, isList = true, defaultValue = "list(0)")
  @Parameter(name = "y", desc = "List of y co-ordinates of stimuli (degrees).", className = Double.class, min = -90, max = 90, isList = true, defaultValue = "list(0)")
  @Parameter(name = "sx", desc = "List of diameters along major axis of ellipse (degrees).", className = Double.class, min = 0, max = 180, isList = true, defaultValue = "list(1.72)")
  @Parameter(name = "sy", desc = "List of diameters along minor axis of ellipse (degrees).", className = Double.class, min = 0, max = 180, isList = true, defaultValue = "list(1.72)")
  @Parameter(name = "lum", desc = "List of stimuli luminances (cd/m^2).", className = Double.class, isList = true, min = 0, max = 3183.099, defaultValue = "list(20)")
  @Parameter(name = "color", desc = "List of stimuli colors.", className = Double[][].class, isList = true, defaultValue = "list(list(1, 1, 1))")
  @Parameter(name = "rotation", desc = "List of angles of rotation of stimuli (degrees). Only useful if sx != sy specified.", className = Double.class, min = 0, max = 360, isList = true, optional = true)
  @Parameter(name = "contrast", desc = "List of stimulus contrasts (from 0 to 1).", className = Double.class, min = 0, max = 1, isList = true, optional = true)
  @Parameter(name = "defocus", desc = "List of defocus values in Diopters for stimulus post-processing.", className = Double.class, min = 0, max = 1, isList = true, optional = true)
  @Parameter(name = "frequency", desc = "List of frequencies (in cycles per degrees) for generation of spatial patterns. Only useful if type != FLAT", className = Double.class, min = 0, max = 1000, isList = true, optional = true)
  @Parameter(name = "patRotation", desc = "List of angles of rotation of stimuli (degrees). Only useful if type != FLAT", className = Double.class, min = 0, max = 360, isList = true, optional = true)
  @Parameter(name = "t", desc = "List of stimuli presentation times (ms).", className = Double.class, min = 0, isList = true, defaultValue = "list(200)")
  @Parameter(name = "w", desc = "List of stimuli response windows (ms).", className = Double.class, min = 0, isList = true, defaultValue = "list(1500)")
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error messages from PhoneHMD.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.seen", desc = "true if seen, false if not.", className = Boolean.class)
  @ReturnMsg(name = "msg.time", desc = "Response time from stimulus onset if button pressed, -1 otherwise (ms).", className = Double.class, min = -1)
  @ReturnMsg(name = "msg.eyex", desc = "x co-ordinates of pupil at times eyet (degrees).", className = Double.class, isList = true)
  @ReturnMsg(name = "msg.eyey", desc = "y co-ordinates of pupil at times eyet (degrees).", className = Double.class, isList = true)
  @ReturnMsg(name = "msg.eyed", desc = "Diameter of pupil at times eyet (degrees).", className = Double.class, isList = true)
  @ReturnMsg(name = "msg.eyet", desc = "Time of (eyex,eyey) pupil relative to stimulus onset t=0 (ms).", className = Double.class, isList = true)
  @ReturnMsg(name = "msg.jovp", desc = "Any JOVP-specific messages that the JOVP sent back.")
  public MessageProcessor.Packet present(HashMap<String, Object> args) {
    return new MessageProcessor.Packet("");
  }

  /**
   * opiClose: Close OPI connection
   * 
   * @param args pairs of argument name and value
   *
   * @return A JSON object with return messages
   *
   * @since 0.0.1
   */
  @ReturnMsg(name = "error", desc = "Empty string for all good, else error messages from PhoneHMD.")
  @ReturnMsg(name = "msg", desc = "JSON Object with all of the other fields described in @ReturnMsg except 'error'.")
  @ReturnMsg(name = "msg.jovp", desc = "Any messages that the JOVP sent back.")
  public MessageProcessor.Packet close() {
    setIsInitialised(false);
    return new MessageProcessor.Packet(true, "");
  }

}

File: opi-rgen/src/main/OPI/R/phoneHMD.r
# Open Perimetry Interface implementation for PhoneHMD
#
# Copyright [2022] [Andrew Turpin & Ivan Marin-Franch]
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require(rjson)

    # environment for this machine in R
if (exists(".opi_env") && !exists("PhoneHMD", where = .opi_env))
    assign("PhoneHMD", new.env(), envir = .opi_env)

#' Implementation of opiInitialise for the PhoneHMD machine.
#'
#' This is for internal use only. Use [opiInitialise()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param ip IP Address of the perimeter.
#' @param port TCP port of the perimeter.
#' @param ip_Monitor IP Address of the OPI JOVP server.
#' @param port_Monitor TCP port of the OPI JOVP server.
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error messages from PhoneHMD.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$jovp Any messages that the JOVP sent back.
#'
#' @examples
#' chooseOpi("PhoneHMD")
#' result <- opiInitialise(ip = "192.126.0.1", port = 50000,
#'                         ip_Monitor = "localhost", port_Monitor = 50001)
#'
#' @seealso [opiInitialise()]
#'
opiInitialise_for_PhoneHMD <- function(ip = NULL, port = NULL, ip_Monitor = NULL, port_Monitor = NULL) {
    .opi_env$PhoneHMD$socket <<- open_socket(ip_Monitor, port_Monitor)
    msg <- list(ip = ip, port = port, ip_Monitor = ip_Monitor, port_Monitor = port_Monitor);
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$PhoneHMD$socket)

    res <- rjson::fromJSON(readLines(.opi_env$PhoneHMD$socket, n=1))
    return(res)
}

#' Implementation of opiPresent for the PhoneHMD machine.
#'
#' This is for internal use only. Use [opiPresent()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param eye Eye to test.
#' @param shape Stimulus shape.
#' @param type Stimulus type.
#' @param x List of x co-ordinates of stimuli (degrees).
#' @param y List of y co-ordinates of stimuli (degrees).
#' @param sx List of diameters along major axis of ellipse (degrees).
#' @param sy List of diameters along minor axis of ellipse (degrees).
#' @param lum List of stimuli luminances (cd/m^2).
#' @param color List of stimuli colors.
#' @param rotation List of angles of rotation of stimuli (degrees). Only useful
#'                    if sx != sy specified.
#' @param contrast List of stimulus contrasts (from 0 to 1).
#' @param defocus List of defocus values in Diopters for stimulus
#'                   post-processing.
#' @param phase List of phases (in degrees) for generation of spatial patterns.
#'                 Only useful if type != FLAT
#' @param frequency List of frequencies (in cycles per degrees) for generation
#'                     of spatial patterns. Only useful if type != FLAT
#' @param textRotation List of angles of rotation of stimuli (degrees). Only
#'                        useful if type != FLAT
#' @param t List of stimuli presentation times (ms).
#' @param w List of stimuli response windows (ms).
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error messages from PhoneHMD.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$seen true if seen, false if not.
#'    - msg$time Response time from stimulus onset if button pressed, -1
#'                  otherwise (ms).
#'    - msg$eyex x co-ordinates of pupil at times eyet (degrees).
#'    - msg$eyey y co-ordinates of pupil at times eyet (degrees).
#'    - msg$eyed Diameter of pupil at times eyet (degrees).
#'    - msg$eyet Time of (eyex,eyey) pupil relative to stimulus onset t=0 (ms).
#'    - msg$jovp Any JOVP-specific messages that the JOVP sent back.
#'
#' @examples
#' chooseOpi("PhoneHMD")
#' result <- opiPresent(stim = list(eye = list('left'), x = list(0), y = list(0),
#'                      sx = list(1.72), sy = list(1.72), lum = list(20),
#'                      color = list(list(1, 1, 1)), t = list(200), w = list(1500)))
#'
#' @seealso [opiPresent()]
#'
opiPresent_for_PhoneHMD <- function(stim = list(eye = NULL, shape = NULL, type = NULL, x = NULL, y = NULL, sx = NULL, sy = NULL, lum = NULL, color = NULL, rotation = NULL, contrast = NULL, defocus = NULL, phase = NULL, frequency = NULL, textRotation = NULL, t = NULL, w = NULL)) {
if(!exists(".opi_env$PhoneHMD$socket") || is.null(.opi_env$PhoneHMD$socket))
    stop("Cannot call opiPresent without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list(eye = stim$eye, shape = stim$shape, type = stim$type, x = stim$x, y = stim$y, sx = stim$sx, sy = stim$sy, lum = stim$lum, color = stim$color, rotation = stim$rotation, contrast = stim$contrast, defocus = stim$defocus, phase = stim$phase, frequency = stim$frequency, textRotation = stim$textRotation, t = stim$t, w = stim$w);
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$PhoneHMD$socket)

    res <- rjson::fromJSON(readLines(.opi_env$PhoneHMD$socket, n=1))
    return(res)
}

#' Implementation of opiSetup for the PhoneHMD machine.
#'
#' This is for internal use only. Use [opiSetup()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param eye Eye to set.
#' @param bgLum Background color for eye.
#' @param bgCol Background color for eye.
#' @param fixType Fixation target type for eye.
#' @param fixLum Fixation target luminance for eye.
#' @param fixCol Fixation target color for eye.
#' @param fixCx x-coordinate of fixation target (degrees).
#' @param fixCy y-coordinate of fixation target (degrees).
#' @param fixSx diameter along major axis of ellipse (degrees).
#' @param fixSy diameter along minor axis of ellipse (degrees).
#' @param fixRotation Angles of rotation of fixation target (degrees). Only
#'                       useful if sx != sy specified.
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error messages from PhoneHMD.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$jovp Any messages that the JOVP sent back.
#'
#' @examples
#' chooseOpi("PhoneHMD")
#' result <- opiSetup(settings = list(eye = "both", bgLum = 10, bgCol = list(1, 1, 1),
#'                    fixType = "maltese", fixLum = 20, fixCol = list(0, 1, 0),
#'                    fixCx = 0, fixCy = 0, fixSx = 1, fixSy = 1, fixRotation = 0))
#'
#' @seealso [opiSetup()]
#'
opiSetup_for_PhoneHMD <- function(settings = list(eye = NULL, bgLum = NULL, bgCol = NULL, fixType = NULL, fixLum = NULL, fixCol = NULL, fixCx = NULL, fixCy = NULL, fixSx = NULL, fixSy = NULL, fixRotation = NULL)) {
if(!exists(".opi_env$PhoneHMD$socket") || is.null(.opi_env$PhoneHMD$socket))
    stop("Cannot call opiSetup without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list(eye = settings$eye, bgLum = settings$bgLum, bgCol = settings$bgCol, fixType = settings$fixType, fixLum = settings$fixLum, fixCol = settings$fixCol, fixCx = settings$fixCx, fixCy = settings$fixCy, fixSx = settings$fixSx, fixSy = settings$fixSy, fixRotation = settings$fixRotation);
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$PhoneHMD$socket)

    res <- rjson::fromJSON(readLines(.opi_env$PhoneHMD$socket, n=1))
    return(res)
}

#' Implementation of opiClose for the PhoneHMD machine.
#'
#' This is for internal use only. Use [opiClose()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#'
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error messages from PhoneHMD.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$jovp Any messages that the JOVP sent back.
#'
#' @examples
#' chooseOpi("PhoneHMD")
#' result <- opiClose()
#'
#' @seealso [opiClose()]
#'
opiClose_for_PhoneHMD <- function() {
if(!exists(".opi_env$PhoneHMD$socket") || is.null(.opi_env$PhoneHMD$socket))
    stop("Cannot call opiClose without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list();
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$PhoneHMD$socket)

    res <- rjson::fromJSON(readLines(.opi_env$PhoneHMD$socket, n=1))
    return(res)
}

#' Implementation of opiQueryDevice for the PhoneHMD machine.
#'
#' This is for internal use only. Use [opiQueryDevice()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#'
#'
#' @return a list contianing:
#'  * error Empty string for all good, else error message.
#'  * msg JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - msg$jovp Any messages that the JOVP sent back.
#'    - msg$isTracking 0 eye tracking is off, any other value it is on.
#'    - msg$isCalibrated 0 eye tracking has not been calibrated, any other value
#'                          it has.
#'
#' @examples
#' chooseOpi("PhoneHMD")
#' result <- opiQueryDevice()
#'
#' @seealso [opiQueryDevice()]
#'
opiQueryDevice_for_PhoneHMD <- function() {
if(!exists(".opi_env$PhoneHMD$socket") || is.null(.opi_env$PhoneHMD$socket))
    stop("Cannot call opiQueryDevice without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list();
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$PhoneHMD$socket)

    res <- rjson::fromJSON(readLines(.opi_env$PhoneHMD$socket, n=1))
    return(res)
}


#' Set background color and luminance in both eyes.
#' Deprecated for ImoVifa and replaced with [opiSetup()].
#' @usage NULL
#' @seealso [opiSetup()]
opiSetBackground_for_ImoVifa <- function(lum, color, ...) {return("Deprecated")}


